Fix several say/think bugs

paulkaplan opened the pull request:
### Resolves

_What Github issue does this resolve (please include link)?_

Resolves https://github.com/LLK/scratch-vm/issues/728

@griffpatch thanks very much for reporting those issues. This PR contains commits to fix that memory leak with event listeners, fixes up the show/hide timeout through green flag clicks and fixes the moving bug. Tiny fixes individually, but together led to some very strange behavior.

Thank you for the thorough repro instructions, that really helps with these bugs.  

For the reviewer, this is one you want to review one commit at a time. The commit description contains more info for each fix.

paulkaplan committed:
Fix event listeners not being cleaned up

Eventually if you hit the green flag a million times with a "green flag
=> say" stack, you'd get an event emitter memory leak error.

paulkaplan committed:
Stop hide timeouts from sticking around through green flag/stop.

paulkaplan committed:
Fix new bubbles not moving with their owning sprite.

This was the original issue #728 that spurred these changes.

rschamp reviewed (changes requested):
Discussed offline, there's still some "unique" visibility behavior, so @paulkaplan is going to look into refactoring the handling of the `visible` state.

paulkaplan committed:
Remove "visible" state flag and use target visibility directly.

This fixes several other bugs like "saying" from a hidden sprite.

paulkaplan committed:
Add back the visibility optimization in a simpler way.

paulkaplan commented:
@rschamp thanks for talking through this. I removed the `visible` state, but then realized what it was originally for: it is an optimization specifically targeted at making "typewriter" style projects work without "flashing" of the bubble by removing it and recreating it. 

I added that back in in the last commit, but what I think is a more stable and sane way. The optimization is only used if the skin has already been initialized, so the relevant code is confined to the single "update" branch of the `renderBubble` function.

cwillisf commented on src/blocks/scratch3_looks.js:
```diff
@@ -6,9 +6,10 @@ const RenderedTarget = require('../sprites/rendered-target');
  * @property {Boolean} onSpriteRight - tracks whether the bubble is right or left of the sprite.
  * @property {?int} drawableId - the ID of the associated bubble Drawable, null if none.
+ * @property {Boolean} drawableVisible - if drawable has been hidden by blank text.
+ *      See _renderBubble for explanation of this optimization.
```
Based on the description here it sounds like `drawableVisible` being true means that the Drawable is hidden. Based on the name of the property and the logic in `_renderBubble` it looks like `drawableVisible` being true means that the Drawable is visible. Unless I'm misunderstanding something, could you adjust for consistency? (Maybe just `false if drawable has been hidden by blank text.`?)

paulkaplan replied:
Oh good catch, sorry I went back and forth adjusting these property names about a dozen times, must have got messed up in the shuffle. Fixed in c14155c

cwillisf reviewed (changes requested):
One nitpick / misunderstanding, but overall LGTM!

paulkaplan committed:
Fix comment on drawableVisible property
